File: GFSSOC '15 Fall S4 - Mostly Talking.py
# 50/100 TLE, PYTHON IS TOO SLOW, CHECK C++ CODE
# shortest path from 1 to N, you are given some special edges, and you can only take 1 special edge in total

import sys
from heapq import heappop, heappush

input = sys.stdin.readline

N, M = map(int, input().split())

graph = [[] for _ in range(N + 1)]  # (destination, distance, special)
for _ in range(M):
    a, b, d = map(int, input().split())
    graph[a].append((b, d, False))

S = int(input())
for _ in range(S):
    a, b, d = map(int, input().split())
    graph[a].append((b, d, True))

inf = 1 << 50

# dist[i][t/f] is the shortest path to node i passing or without passing through a special node
dist = [[inf, inf] for _ in range(N + 1)]
dist[1] = [0, 0]

pq = [(0, 1, False)]
while pq:
    d, cur, special = heappop(pq)
    if cur == N:
        print(d)
        sys.exit()

    for adj, adj_dist, adj_special in graph[cur]:
        new_dist = adj_dist + d
        new_special = special + adj_special
        if new_special > 1:  # can't take more than 1 special path
            continue

        if dist[adj][new_special] > new_dist:
            dist[adj][new_special] = new_dist
            heappush(pq, (new_dist, adj, new_special))

print(-1)

File: GFSSOC '15 Fall S4 - Mostly Talking.cpp
#include <bits/stdc++.h>
using namespace std;

const long long INF = 1LL << 50;
typedef pair<long long, pair<int, bool>> plib;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M;
    cin >> N >> M;

    vector<vector<pair<int, pair<long long, bool>>>> graph(N + 1);  // (destination, distance, special)
    for(int i = 0; i < M; i++) {
        int a, b;
        long long d;
        cin >> a >> b >> d;
        graph[a].push_back({b, {d, false}});
    }

    int S;
    cin >> S;
    for(int i = 0; i < S; i++) {
        int a, b;
        long long d;
        cin >> a >> b >> d;
        graph[a].push_back({b, {d, true}});
    }

    vector<vector<long long>> dist(N + 1, vector<long long>(2, INF));
    dist[1][0] = 0;

    priority_queue<plib, vector<plib>, greater<plib>> pq;
    pq.push({0, {1, false}});

    while(!pq.empty()) {
        long long d = pq.top().first;
        int cur = pq.top().second.first;
        bool special = pq.top().second.second;
        pq.pop();

        if(cur == N) {
            cout << d << "\n";
            return 0;
        }

        for(auto &edge : graph[cur]) {
            int adj = edge.first;
            long long adj_dist = edge.second.first;
            bool adj_special = edge.second.second;

            long long new_dist = adj_dist + d;
            int new_special = special + adj_special;

            if(new_special > 1)  // can't take more than 1 special path
                continue;

            if(dist[adj][new_special] > new_dist) {
                dist[adj][new_special] = new_dist;
                pq.push({new_dist, {adj, new_special}});
            }
        }
    }

    cout << -1 << "\n";
    return 0;
}
